addClass() doesn't appear to have an effect

I did the following:

1. Download Jet 0.3.1 from the website
2. bin/start.sh
3. execute the client code:

```java
public class Main {
    public static void main(String[] args) throws Exception {
        try {
            ClientConfig conf = new ClientConfig();
            conf.setCredentials(new UsernamePasswordCredentials("jet", "jet-pass"));
            JetInstance jet = Jet.newJetClient(conf);
            IStreamMap<Object, Object> events = jet.getMap("events");
            IStreamMap<Object, Object> newevents = jet.getMap("newevents");
            events.clear();
            newevents.clear();
            for (int i = 0; i < 1000; i++) {
                events.put(i, new Event(System.nanoTime(), "key" + i, "value"));
            }
            JobConfig jconf = new JobConfig().addClass(Event.class);
            jet.newJob(buildDag(), jconf).execute().get();
            System.out.println("Output map size: " + newevents.size());
        } finally {
            Jet.shutdownAll();
        }
    }

    private static DAG buildDag() {
        DAG dag = new DAG();
        Vertex source = dag.newVertex("source", readMap("events"));
        Vertex sink = dag.newVertex("sink", writeMap("newevents"));
        dag.edge(between(source, sink));
        return dag;
    }
}

public class Event implements Serializable {
    long timestamp;
    String key;
    String value;

    Event(long timestamp, String key, String value) {
        this.timestamp = timestamp;
        this.key = key;
        this.value = value;
    }
}
```

The outcome is:

```
Job execution failed with exception
Apr 17, 2017 6:14:08 PM com.hazelcast.core.LifecycleService
INFO: hz.client_0 [dev] [0.3.1] [3.8] HazelcastClient 3.8 (20170217 - d7998b4) is CLIENT_CONNECTED
java.util.concurrent.ExecutionException: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{vertex=sink, processor=com.hazelcast.jet.impl.connector.WriteIMapP@79e48f8e}: com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: test.Event
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:87)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:31)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155)
	at com.hazelcast.jet.impl.JetClientInstanceImpl$ExecutionFuture.get(JetClientInstanceImpl.java:177)
	at com.hazelcast.jet.impl.JetClientInstanceImpl$ExecutionFuture.get(JetClientInstanceImpl.java:132)
	at test.Main.main(Main.java:31)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)Job execution failed with exception

	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: com.hazelcast.jet.JetException: Exception in ProcessorTasklet{vertex=sink, processor=com.hazelcast.jet.impl.connector.WriteIMapP@79e48f8e}: com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: test.Event
	at com.hazelcast.jet.impl.execution.ExecutionService$BlockingWorker.run(ExecutionService.java:165)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:114)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:243)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:114)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$1.run(AbstractInvocationFuture.java:243)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
	at ------ submitted from ------.(Unknown Source)
	at com.hazelcast.client.spi.impl.ClientInvocationFuture.resolveAndThrowIfException(ClientInvocationFuture.java:74)
	... 10 more
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: test.Event
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:224)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:184)
	at com.hazelcast.query.impl.CachedQueryEntry.getValue(CachedQueryEntry.java:70)
	at com.hazelcast.map.impl.proxy.MapProxySupport.putAllInternal(MapProxySupport.java:796)
	at com.hazelcast.map.impl.proxy.MapProxyImpl.putAll(MapProxyImpl.java:377)
	at com.hazelcast.jet.impl.connector.WriteIMapP.flush(WriteIMapP.java:68)
	at com.hazelcast.jet.impl.connector.WriteIMapP.process(WriteIMapP.java:52)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.tryProcessInbox(ProcessorTasklet.java:148)
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:99)
	at com.hazelcast.jet.impl.execution.ExecutionService$BlockingWorker.run(ExecutionService.java:155)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: test.Event
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:146)
	at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:120)
	at com.hazelcast.nio.IOUtil$ClassLoaderAwareObjectInputStream.resolveClass(IOUtil.java:522)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1613)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1518)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1774)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:371)
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:219)
	... 15 more
Output map size: 0
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ExecutionContext.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ExecutionContext.java
@@ -17,6 +17,7 @@
 package com.hazelcast.jet.impl.execution;
 
 import com.hazelcast.jet.ProcessorSupplier;
+import com.hazelcast.jet.impl.JetService;
 import com.hazelcast.jet.impl.execution.init.ExecutionPlan;
 import com.hazelcast.nio.Address;
 import com.hazelcast.nio.BufferObjectDataInput;
@@ -58,7 +59,9 @@ public class ExecutionContext {
     }
 
     public CompletionStage<Void> execute(Consumer<CompletionStage<Void>> doneCallback) {
-        executionCompletionStage = execService.execute(tasklets, doneCallback);
+        JetService service = nodeEngine.getService(JetService.SERVICE_NAME);
+        ClassLoader cl = service.getClassLoader(executionId);
+        executionCompletionStage = execService.execute(tasklets, doneCallback, cl);
         executionCompletionStage.whenComplete((r, e) -> tasklets.clear());
         return executionCompletionStage;
     }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ExecutionService.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ExecutionService.java
@@ -40,6 +40,7 @@ import java.util.concurrent.atomic.AtomicReference;
 import java.util.concurrent.locks.LockSupport;
 import java.util.function.Consumer;
 
+import static java.lang.Thread.currentThread;
 import static java.util.concurrent.Executors.newCachedThreadPool;
 import static java.util.concurrent.TimeUnit.MICROSECONDS;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
@@ -50,6 +51,7 @@ public class ExecutionService {
 
     static final IdleStrategy IDLER =
             new BackoffIdleStrategy(0, 0, MICROSECONDS.toNanos(1), MILLISECONDS.toNanos(1));
+
     private final ExecutorService blockingTaskletExecutor = newCachedThreadPool(new BlockingTaskThreadFactory());
     private final CooperativeWorker[] workers;
     private final Thread[] threads;
@@ -67,14 +69,17 @@ public class ExecutionService {
      * @return instance of {@code java.util.concurrent.CompletableFuture}
      */
     public CompletionStage<Void> execute(
-            List<? extends Tasklet> tasklets, @Nonnull Consumer<CompletionStage<Void>> doneCallback) {
+            @Nonnull List<? extends Tasklet> tasklets,
+            @Nonnull Consumer<CompletionStage<Void>> doneCallback,
+            @Nonnull ClassLoader jobClassLoader
+    ) {
         ensureStillRunning();
         final JobFuture jobFuture = new JobFuture(tasklets.size(), doneCallback);
         try {
             final Map<Boolean, List<Tasklet>> byCooperation =
                     tasklets.stream().collect(partitioningBy(Tasklet::isCooperative));
-            submitCooperativeTasklets(jobFuture, byCooperation.get(true));
-            submitBlockingTasklets(jobFuture, byCooperation.get(false));
+            submitCooperativeTasklets(jobFuture, jobClassLoader, byCooperation.get(true));
+            submitBlockingTasklets(jobFuture, jobClassLoader, byCooperation.get(false));
         } catch (Throwable t) {
             jobFuture.completeExceptionally(t);
             doneCallback.accept(jobFuture);
@@ -99,21 +104,21 @@ public class ExecutionService {
         }
     }
 
-    private void submitBlockingTasklets(JobFuture jobFuture, List<Tasklet> tasklets) {
+    private void submitBlockingTasklets(JobFuture jobFuture, ClassLoader jobClassLoader, List<Tasklet> tasklets) {
         jobFuture.blockingFutures = tasklets.stream()
-                                            .map(t -> new BlockingWorker(new TaskletTracker(t, jobFuture)))
+                                            .map(t -> new BlockingWorker(new TaskletTracker(t, jobFuture, jobClassLoader)))
                                             .map(blockingTaskletExecutor::submit)
                                             .collect(toList());
     }
 
-    private void submitCooperativeTasklets(JobFuture jobFuture, List<Tasklet> tasklets) {
+    private void submitCooperativeTasklets(JobFuture jobFuture, ClassLoader jobClassLoader, List<Tasklet> tasklets) {
         ensureThreadsStarted();
         final List<TaskletTracker>[] trackersByThread = new List[workers.length];
         Arrays.setAll(trackersByThread, i -> new ArrayList());
         int i = 0;
         for (Tasklet t : tasklets) {
             t.init(jobFuture);
-            trackersByThread[i++ % trackersByThread.length].add(new TaskletTracker(t, jobFuture));
+            trackersByThread[i++ % trackersByThread.length].add(new TaskletTracker(t, jobFuture, jobClassLoader));
         }
         for (i = 0; i < trackersByThread.length; i++) {
             workers[i].trackers.addAll(trackersByThread[i]);
@@ -140,7 +145,9 @@ public class ExecutionService {
 
         @Override
         public void run() {
+            final ClassLoader clBackup = currentThread().getContextClassLoader();
             final Tasklet t = tracker.tasklet;
+            currentThread().setContextClassLoader(tracker.jobClassLoader);
             try {
                 t.init(tracker.jobFuture);
                 long idleCount = 0;
@@ -157,6 +164,7 @@ public class ExecutionService {
                 logger.warning("Exception in " + t, e);
                 tracker.jobFuture.completeExceptionally(new JetException("Exception in " + t + ": " + e, e));
             } finally {
+                currentThread().setContextClassLoader(clBackup);
                 tracker.jobFuture.taskletDone();
             }
         }
@@ -174,6 +182,8 @@ public class ExecutionService {
 
         @Override
         public void run() {
+            final Thread thread = currentThread();
+            final ClassLoader clBackup = thread.getContextClassLoader();
             long idleCount = 0;
             while (!isShutdown) {
                 boolean madeProgress = false;
@@ -186,6 +196,7 @@ public class ExecutionService {
                         continue;
                     }
                     try {
+                        thread.setContextClassLoader(t.jobClassLoader);
                         final ProgressState result = t.tasklet.call();
                         if (result.isDone()) {
                             dismissTasklet(t);
@@ -203,6 +214,7 @@ public class ExecutionService {
                 if (madeProgress) {
                     idleCount = 0;
                 } else {
+                    thread.setContextClassLoader(clBackup);
                     IDLER.idle(++idleCount);
                 }
             }
@@ -243,11 +255,13 @@ public class ExecutionService {
     private static final class TaskletTracker {
         final Tasklet tasklet;
         final JobFuture jobFuture;
+        final ClassLoader jobClassLoader;
         final AtomicReference<CooperativeWorker> stealingWorker = new AtomicReference<>();
 
-        TaskletTracker(Tasklet tasklet, JobFuture jobFuture) {
+        TaskletTracker(Tasklet tasklet, JobFuture jobFuture, ClassLoader jobClassLoader) {
             this.tasklet = tasklet;
             this.jobFuture = jobFuture;
+            this.jobClassLoader = jobClassLoader;
         }
     }
 

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/execution/ExecutionServiceTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/execution/ExecutionServiceTest.java
@@ -42,10 +42,11 @@ import java.util.concurrent.ExecutionException;
 import java.util.function.Consumer;
 import java.util.stream.Stream;
 
+import static com.hazelcast.jet.DistributedFunctions.noopConsumer;
+import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
 import static com.hazelcast.jet.impl.util.ProgressState.DONE;
 import static com.hazelcast.jet.impl.util.ProgressState.MADE_PROGRESS;
 import static com.hazelcast.jet.impl.util.ProgressState.NO_PROGRESS;
-import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
 import static java.util.Arrays.asList;
 import static java.util.Collections.singletonList;
 import static java.util.stream.Collectors.toList;
@@ -59,12 +60,13 @@ import static org.mockito.Mockito.mock;
 @RunWith(HazelcastSerialClassRunner.class)
 public class ExecutionServiceTest extends JetTestSupport {
 
-    private final Consumer<CompletionStage<Void>> doneCallback = x -> {};
+    private final Consumer<CompletionStage<Void>> doneCallback = noopConsumer();
 
     @Rule
     public final ExpectedException exceptionRule = ExpectedException.none();
 
-    ExecutionService es;
+    private ExecutionService es;
+    private ClassLoader classLoaderMock;
 
     @Before
     public void before() {
@@ -75,6 +77,7 @@ public class ExecutionServiceTest extends JetTestSupport {
         Mockito.when(hzMock.getLoggingService()).thenReturn(loggingService);
         Mockito.when(loggingService.getLogger(Mockito.<Class>any())).thenReturn(mockLogger);
         es = new ExecutionService(hzMock, 4);
+        classLoaderMock = mock(ClassLoader.class);
     }
 
     @Test
@@ -83,7 +86,7 @@ public class ExecutionServiceTest extends JetTestSupport {
         final MockTasklet t = new MockTasklet().blocking();
 
         // When
-        es.execute(singletonList(t), doneCallback).toCompletableFuture().join();
+        es.execute(singletonList(t), doneCallback, classLoaderMock).toCompletableFuture().join();
 
         // Then
         t.assertDone();
@@ -95,7 +98,7 @@ public class ExecutionServiceTest extends JetTestSupport {
         final MockTasklet t = new MockTasklet();
 
         // When
-        es.execute(singletonList(t), doneCallback).toCompletableFuture().join();
+        es.execute(singletonList(t), doneCallback, classLoaderMock).toCompletableFuture().join();
 
         // Then
         t.assertDone();
@@ -107,7 +110,7 @@ public class ExecutionServiceTest extends JetTestSupport {
         final MockTasklet t = new MockTasklet().initFails();
 
         // When
-        es.execute(singletonList(t), doneCallback).toCompletableFuture().join();
+        es.execute(singletonList(t), doneCallback, classLoaderMock).toCompletableFuture().join();
 
         // Then
         t.assertDone();
@@ -119,7 +122,7 @@ public class ExecutionServiceTest extends JetTestSupport {
         final MockTasklet t = new MockTasklet().blocking().initFails();
 
         // When - Then
-        es.execute(singletonList(t), doneCallback).toCompletableFuture().join();
+        es.execute(singletonList(t), doneCallback, classLoaderMock).toCompletableFuture().join();
     }
 
     @Test(expected = CompletionException.class)
@@ -128,7 +131,7 @@ public class ExecutionServiceTest extends JetTestSupport {
         final MockTasklet t = new MockTasklet().callFails();
 
         // When - Then
-        es.execute(singletonList(t), doneCallback).toCompletableFuture().join();
+        es.execute(singletonList(t), doneCallback, classLoaderMock).toCompletableFuture().join();
     }
 
     @Test(expected = CompletionException.class)
@@ -137,21 +140,21 @@ public class ExecutionServiceTest extends JetTestSupport {
         final MockTasklet t = new MockTasklet().blocking().callFails();
 
         // When - Then
-        es.execute(singletonList(t), doneCallback).toCompletableFuture().join();
+        es.execute(singletonList(t), doneCallback, classLoaderMock).toCompletableFuture().join();
     }
 
     @Test
     public void when_shutdown_then_submitFails() {
         // Given
-        es.execute(singletonList(new MockTasklet()), doneCallback);
-        es.execute(singletonList(new MockTasklet()), doneCallback);
+        es.execute(singletonList(new MockTasklet()), doneCallback, classLoaderMock);
+        es.execute(singletonList(new MockTasklet()), doneCallback, classLoaderMock);
 
         // When
         es.shutdown();
 
         // Then
         exceptionRule.expect(IllegalStateException.class);
-        es.execute(singletonList(new MockTasklet()), doneCallback);
+        es.execute(singletonList(new MockTasklet()), doneCallback, classLoaderMock);
     }
 
     @Test
@@ -162,7 +165,7 @@ public class ExecutionServiceTest extends JetTestSupport {
                 new MockTasklet().callsBeforeDone(10));
 
         // When
-        es.execute(tasklets, doneCallback).toCompletableFuture().join();
+        es.execute(tasklets, doneCallback, classLoaderMock).toCompletableFuture().join();
 
         // Then
         tasklets.forEach(MockTasklet::assertDone);
@@ -176,7 +179,7 @@ public class ExecutionServiceTest extends JetTestSupport {
                       .limit(100).collect(toList());
 
         // When
-        es.execute(tasklets, doneCallback).toCompletableFuture().join();
+        es.execute(tasklets, doneCallback, classLoaderMock).toCompletableFuture().join();
 
         // Then
         tasklets.forEach(MockTasklet::assertDone);
@@ -190,7 +193,7 @@ public class ExecutionServiceTest extends JetTestSupport {
                       .limit(100).collect(toList());
 
         // When
-        CompletableFuture<Void> future = es.execute(tasklets, doneCallback).toCompletableFuture();
+        CompletableFuture<Void> future = es.execute(tasklets, doneCallback, classLoaderMock).toCompletableFuture();
         future.cancel(true);
 
         // Then
@@ -208,7 +211,7 @@ public class ExecutionServiceTest extends JetTestSupport {
                       .limit(100).collect(toList());
 
         // When
-        CompletableFuture<Void> future = es.execute(tasklets, doneCallback).toCompletableFuture();
+        CompletableFuture<Void> future = es.execute(tasklets, doneCallback, classLoaderMock).toCompletableFuture();
         future.cancel(true);
 
         // Then
@@ -226,7 +229,7 @@ public class ExecutionServiceTest extends JetTestSupport {
                       .limit(100).collect(toList());
 
         // When
-        CompletableFuture<Void> future = es.execute(tasklets, doneCallback).toCompletableFuture();
+        CompletableFuture<Void> future = es.execute(tasklets, doneCallback, classLoaderMock).toCompletableFuture();
         future.cancel(true);
 
         // Then
@@ -246,7 +249,7 @@ public class ExecutionServiceTest extends JetTestSupport {
         CompletableFuture<Void> doneFuture = new CompletableFuture<Void>();
 
         // When
-        CompletableFuture<Void> future = es.execute(tasklets, f -> doneFuture.complete(null))
+        CompletableFuture<Void> future = es.execute(tasklets, f -> doneFuture.complete(null), classLoaderMock)
                                            .toCompletableFuture();
 
         future.cancel(true);

